src/dataflow/src/sink/kafka.rs

```diff
@@ -566,17 +690,185 @@ impl KafkaSinkState {
+            // therefore cannot simply take the last offset from the back and expect an
+            // END message there. With a transactional producer, the OffsetTail(1) will
+            // not point to an END message but a control message. With aborted
+            // transactions, there might even be a lot of garbage at the end of the
+            // topic or in between.
+
+            let mut tps = TopicPartitionList::new();
+            tps.add_partition(consistency_topic, partition);
+            tps.set_partition_offset(consistency_topic, partition, Offset::Beginning)?;
+
+            consumer.assign(&tps).with_context(|| {
+                format!(
+                    "Error seeking in consistency topic {}:{}",
+                    consistency_topic, partition
+                )
+            })?;
+
+            let (lo, hi) = consumer
+                .fetch_watermarks(consistency_topic, 0, timeout)
+                .map_err(|e| {
+                    anyhow!(
+                        "Failed to fetch metadata while reading from consistency topic: {}",
+                        e
+                    )
+                })?;
+
+            // Empty topic.  Return early to avoid unnecesasry call to kafka below.
+            if hi == 0 {
+                return Ok(None);
+            }
+
+            let mut latest_ts = None;
+            let mut latest_offset = None;
+            while let Some((message, offset)) = get_next_message(&mut consumer, timeout)? {
+                debug_assert!(offset >= latest_offset.unwrap_or(0));
+                latest_offset = Some(offset);
+
+                if let Some(ts) = maybe_decode_consistency_end_record(&message, consistency_topic)?
+                {
+                    if ts >= latest_ts.unwrap_or(0) {
+                        latest_ts = Some(ts);
+                    }
+                }
+            }
+
+            // Topic not empty but we couldn't read any messages.  We don't expect this to happen but we
+            // have no reason to rely on kafka not inserting any internal messages at the beginning.
+            if latest_offset.is_none() {
+                tracing::debug!(
+                    "unable to read any messages from non-empty topic {}:{}, lo/hi: {}/{}",
+                    consistency_topic,
+                    partition,
+                    lo,
+                    hi
+                );
+            }
+            Ok(latest_ts)
+        }
+
+        // There may be arbitrary messages in this topic that we cannot decode.  We only
+        // return an error when we know we've found an END message but cannot decode it.
+        fn maybe_decode_consistency_end_record(
+            bytes: &[u8],
+            consistency_topic: &str,
+        ) -> Result<Option<Timestamp>, anyhow::Error> {
+            // The first 5 bytes are reserved for the schema id/schema registry information
+            let mut bytes = &bytes[5..];
```

Comment:
If bytes is malformed this will panic, can you add a length check and return an error if it's < 5?